File: prototype/rails/src/test/domain/auth/magic_link_status_test.rb
require "test_helper"

module Domain
  module Auth
    class MagicLinkStatusTest < ActiveSupport::TestCase
      EXPIRES_AT = Time.utc(2026, 8, 22, 12, 15)

      test "a fresh unconsumed link is usable" do
        assert_equal MagicLinkStatus::USABLE, status_at(Time.utc(2026, 8, 22, 12, 0))
      end

      test "a link expires the moment now reaches the expiry" do
        assert_equal MagicLinkStatus::EXPIRED, status_at(EXPIRES_AT)
      end

      test "a link is expired after the expiry" do
        assert_equal MagicLinkStatus::EXPIRED, status_at(EXPIRES_AT + 1)
      end

      test "a consumed link is consumed" do
        assert_equal MagicLinkStatus::CONSUMED,
          status_at(Time.utc(2026, 8, 22, 12, 6), consumed_at: Time.utc(2026, 8, 22, 12, 5))
      end

      test "consumption outranks expiry" do
        assert_equal MagicLinkStatus::CONSUMED,
          status_at(Time.utc(2026, 8, 22, 13, 0), consumed_at: Time.utc(2026, 8, 22, 12, 5))
      end

      private

      def status_at(now, consumed_at: nil)
        MagicLinkStatus.of(expires_at: EXPIRES_AT, consumed_at: consumed_at, now: now)
      end
    end
  end
end

File: app/src/app/Domain/Auth/MagicLinkStatusTest.php
<?php

declare(strict_types=1);

namespace App\Domain\Auth;

use DateTimeImmutable;

it('derives status from expiry, consumption, and now', function (
    string $expiresAt,
    ?string $consumedAt,
    string $now,
    MagicLinkStatus $expected,
): void {
    $status = MagicLinkStatus::of(
        new DateTimeImmutable($expiresAt),
        $consumedAt === null ? null : new DateTimeImmutable($consumedAt),
        new DateTimeImmutable($now),
    );

    expect($status)->toBe($expected);
})->with([
    'a fresh unconsumed link is usable' => ['2026-08-22 12:15:00', null, '2026-08-22 12:00:00', MagicLinkStatus::Usable],
    'a link is expired once now reaches the expiry' => ['2026-08-22 12:15:00', null, '2026-08-22 12:15:00', MagicLinkStatus::Expired],
    'a link is expired after the expiry' => ['2026-08-22 12:15:00', null, '2026-08-22 12:15:01', MagicLinkStatus::Expired],
    'a consumed link is consumed' => ['2026-08-22 12:15:00', '2026-08-22 12:05:00', '2026-08-22 12:06:00', MagicLinkStatus::Consumed],
    'consumption outranks expiry' => ['2026-08-22 12:15:00', '2026-08-22 12:05:00', '2026-08-22 13:00:00', MagicLinkStatus::Consumed],
]);
